httputils.nim

```diff
@@ -1647,3 +1649,40 @@ proc toInt*(code: HttpCode): int =
     of Http503: 503
     of Http504: 504
     of Http505: 505
+
+proc improveMatch(initialAcceptHeadersMatchIndex: int,
+                  initialAppPreferencesMatchIndex: int,
+                  matchQ: float,
+                  sortedAcceptHeaders: seq[AcceptMediaItem],
+                  sortedAppPreferences: varargs[MediaType]): MediaType =
+  var bestAppPreferencesIndex = initialAppPreferencesMatchIndex
+  for accept in sortedAcceptHeaders[initialAcceptHeadersMatchIndex .. ^1]:
+    if accept.qvalue < matchQ:
+      # If 'q' parameter change, no point continuing
+      break
+    # Searching for better match between lower indexes of application preferences
+    # than current match
+    for index, offer in sortedAppPreferences[0..initialAppPreferencesMatchIndex].pairs:
```

Comment:
This algorithm still doesn't look right. You are selecting the first better match you can find, but that's still not necessarily the best match. Consider the following inputs:

Accept Headers: A, B, C (all with equal weight)
App Preferences: C, B, A

The outer algorithm will match A right away. `improveMatch` will then proceed to search for an improvement and it will discover that B is an improvement and it would return it. This is still not right, because C was the best choice.

To solve this, you have to remove the `break` here and and iterate up to `bestAppPreferencesIndex` in the search loop. Only the `break` statement in the outer loop (i.e. `if accept.qvalue < matchQ`) should be the way to exit the function.

Another problem with the code is that looping with the construct `sortedAppPreferences[0..initialAppPreferencesMatchIndex].pairs` leads to a temporary allocation of a new sequence and a copy of the requested slice (I know, Nim is silly). This overhead is completely unnecessary and the code will be better if you use regular index-based iteration instead:

```nim
for offerIdx in 0 ..< initialAppPreferencesMatchIndex:
  if accept.mediaType == sortedAppPreferences[offerIdx]:
    ...
```

As you may notice, I also used `..<` instead of `..` because your code currently suffers from an off-by-one error.

As an alternative to index-based iteration, you can also use `for index, offer in sortedAppPreferences.toOpenArray(0, initialAppPreferencesMatchIndex - 1)` as this doesn't lead to allocations, but that's quite the mouthful of code :)

----

As a more general remark for the PR, I can see that this functionality is now tested in the Chronos repo, but it makes sense to move the tests here, so future changes to this repo are properly tested by its own CI.